demos/http/http_demo_s3_upload/http_demo_s3_upload.c

```diff
@@ -0,0 +1,626 @@
+
+/**
+ * @brief Connect to HTTP server with reconnection retries.
+ *
+ * @param[out] pNetworkContext The output parameter to return the created network context.
+ *
+ * @return EXIT_FAILURE on failure; EXIT_SUCCESS on successful connection.
+ */
+static int32_t connectToServer( NetworkContext_t * pNetworkContext );
+
+/**
+ * @brief Retrieve the size of the S3 object that is specified in pPath.
+ *
+ * @param[out] pFileSize - The size of the S3 object.
+ * @param[in] pTransportInterface The transport interface for making network calls.
+ * @param[in] pHost The host name of the server.
+ * @param[in] pMethod The HTTP request method.
+ * @param[in] pPath The Request-URI to the objects of interest.
+ */
+static HTTPStatus_t getS3ObjectFileSize( size_t * pFileSize,
+                                         const TransportInterface_t * pTransportInterface,
+                                         const char * pHost,
+                                         const char * pMethod,
+                                         const char * pPath );
+
+/**
+ * @brief Send an HTTP request based on a specified method and path, then
+ * print the response received from the server.
+ *
+ * @param[in] pTransportInterface The transport interface for making network calls.
+ * @param[in] pHost The host name of the server.
+ * @param[in] pMethod The HTTP request method.
+ * @param[in] pPath The Request-URI to the objects of interest.
+ *
+ * @return EXIT_FAILURE on failure; EXIT_SUCCESS on success.
+ */
+static int uploadS3ObjectFile( const TransportInterface_t * pTransportInterface,
+                               const char * pHost,
+                               const char * pMethod,
+                               const char * pPath );
+
+/*-----------------------------------------------------------*/
+
+static int32_t connectToServer( NetworkContext_t * pNetworkContext )
+{
+    int32_t returnStatus = EXIT_FAILURE;
+    HTTPStatus_t httpStatus = HTTP_SUCCESS;
+
+    /* The location of the host address within the pre-signed URL. */
+    const char * pAddress = NULL;
+    /* The length of the host address found in the pre-signed URL. */
+    size_t serverHostLength = 0;
+
+    /* Status returned by OpenSSL transport implementation. */
+    OpensslStatus_t opensslStatus;
+    /* Credentials to establish the TLS connection. */
+    OpensslCredentials_t opensslCredentials;
+    /* Information about the server to send the HTTP requests. */
+    ServerInfo_t serverInfo;
+
+    /* Initialize TLS credentials. */
+    ( void ) memset( &opensslCredentials, 0, sizeof( opensslCredentials ) );
+    opensslCredentials.pRootCaPath = ROOT_CA_CERT_PATH;
+
+    /* ALPN is required when communicating to AWS IoT Core over port 443 through HTTP. */
+    if( HTTPS_PORT == 443 )
+    {
+        opensslCredentials.pAlpnProtos = IOT_CORE_ALPN_PROTOCOL_NAME;
+        opensslCredentials.alpnProtosLen = strlen( IOT_CORE_ALPN_PROTOCOL_NAME );
```

Comment:
We are discouraged from using `strlen` in demos. You are allowed to use it though in unit tests and integration tests. I know it's a hassle, but you can also create a IOT_CORE_ALPN_PROTOCOL_NAME_LENGTH macro.

Reply:
Sounds good -- changed all instances of `strlen`.